expression/builtin_string.go

```diff
@@ -1366,3 +1367,83 @@ func builtinFindInSet(args []types.Datum, _ context.Context) (d types.Datum, err
 	}
 	return
 }
+
+type fieldFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *fieldFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	return &builtinFieldSig{newBaseBuiltinFunc(args, ctx)}, errors.Trace(c.verifyArgs(args))
+}
+
+type builtinFieldSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinFieldSig) eval(row []types.Datum) (types.Datum, error) {
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return types.Datum{}, errors.Trace(err)
+	}
+	return builtinField(args, b.ctx)
+}
+
+// See http://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_field
+// Returns the index (position) of arg0 in the arg1, arg2, arg3, ... list.
+// Returns 0 if arg0 is not found.
+// If arg0 is NULL, the return value is 0 because NULL fails equality comparison with any value.
+// If all arguments are strings, all arguments are compared as strings.
+// If all arguments are numbers, they are compared as numbers.
+// Otherwise, the arguments are compared as double.
+func builtinField(args []types.Datum, ctx context.Context) (d types.Datum, err error) {
+	d.SetInt64(0)
+	if args[0].IsNull() {
+		return
+	}
+	var (
+		pos       int64
+		allString bool
+		newArgs   []types.Datum
+	)
+	allString = true
+	for i := 0; i < len(args) && allString; i++ {
+		switch args[i].Kind() {
+		case types.KindInt64, types.KindUint64, types.KindFloat32, types.KindFloat64, types.KindMysqlDecimal:
+			allString = false
+		}
+	}
+	newArgs, err = argsToSpecifiedType(args, allString, ctx)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	arg0, sc := newArgs[0], ctx.GetSessionVars().StmtCtx
+	for i, curArg := range newArgs[1:] {
+		cmpResult, _ := arg0.CompareDatum(sc, curArg)
+		if cmpResult == 0 {
+			pos = int64(i + 1)
+		}
+	}
+	d.SetInt64(pos)
+	return d, errors.Trace(err)
+}
+
+// argsToSpecifiedType converts the type of all arguments in args into string type if allString is true,
+// otherwise converts them into double type.
+func argsToSpecifiedType(args []types.Datum, allString bool, ctx context.Context) (newArgs []types.Datum, err error) {
```

Comment:
If all numbers are decimal, converting it to double may lose precision, then the result will be wrong.
We should follow MySQL document.